### PR TITLE
core: introduce TypedAttributeConstraint

### PR DESCRIPTION
**Pull request by alexarice**
Experiment in using Constraint variables to grab a type from an attribute, and an example of using this for `arith.constant`.

Unsure if this a good idea or cooking too hard.

Edit: I also couldn't really figure out how `TypedAttribute` worked, and was not sure if the functionality could be reused (I have use cases in mind where the associated type would not be a parameter)

### CODE REVIEW

**Comment by alexarice** [comment]
I also believe both failing tests are actually bugs exposed by this change rather than bugs with this change

**Review by superlopuh — COMMENTED** [review]
On one hand, you're cooking pretty hard indeed, on the other this is great. Probably worth integrating with `TypedAttribute[Attribute]`

**Comment by math-fehr** [comment]
Do you think you could have an inheritance like `WithType` -> `TypedAttribute`?
`WithType` just means your attribute has a canonical type. On the other hand, `TypedAttribute` means that you have a type, that the type is stored in one of the parameters, and that the type is parsed after the `:`. So `TypedAttribute` should be a strict subset of `WithType`.

My goal with `TypedAttribute` was originally to handle all cases that are (or can be) parsed with `:`.
You said you had in mind attributes that do not have types in their parameters, what example do you have in mind?

**Comment by alexarice** [comment]
@math-fehr with the `TypedAttribute` class, is it guaranteed that there is a `type` attribute? Pyright doesn't seem to know this but I found a place where I think it was being runtime enforced

**Comment by alexarice** [comment]
This is the code I was referring to:
https://github.com/xdslproject/xdsl/blob/61bf63244207133d9394260d5a7a4bcb568dcc4e/xdsl/irdl/attributes.py#L163-L181
perhaps this is just checking that the type parameter can be found in the correct place.

I'll rebase this in the meantime

**Comment by alexarice** [comment]
I looked again and `IntegerAttr` seems to be the only `TypedAttribute` in xdsl

**Comment by alexarice** [comment]
Would you expect `DenseIntOrFPElementsAttr` to be a `TypedAttribute`? I'm beginning to think that the functionality of "having an associated type" and "printing and parsing with/without a type" are slightly separate

**Comment by alexarice** [comment]
I want to add that I've been experimenting with traits like this a bit more, and it's all kind of evil as it all depends on what order you "verify" constraints in. Ideally, I feel that the "verify" methods should not solve any constraints and all solving of constraint variables should be done by a separate system which does not depend on the order of constraints, though this would need a bigger refactor.

If anyone has had any ideas on this in general, it would be good to know.

File: xdsl/dialects/arith.py [hunk 1]
```diff
@@ -123,11 +127,21 @@ def __init__(self, flags: None | Sequence[IntegerOverflowFlag] | Literal["none"]
 @irdl_op_definition
 class Constant(IRDLOperation):
     name = "arith.constant"
-    result = result_def(Attribute)
-    value = prop_def(Attribute)
+    _T: ClassVar[AttrConstraint] = VarConstraint("T", AnyAttr())
```
Comment (superlopuh) [inline xdsl/dialects/arith.py:130]:
```suggestion
    _T: ClassVar = VarConstraint("T", AnyAttr())
```

**Comment by math-fehr** [comment]
> Would you expect `DenseIntOrFPElementsAttr` to be a `TypedAttribute`? I'm beginning to think that the functionality of "having an associated type" and "printing and parsing with/without a type" are slightly separate

Sorry I thought I answered this but I forgot to.
If I recall, these are merged in MLIR. In MLIR, all attribute have a `type` field (most of them are just `None` though).
So if you have an associated type, it will be printed/parsed with the colon. That's my recollection at least.
We are free to do however we want, though I would likely say that these should be merged.

**Comment by math-fehr** [comment]
> I want to add that I've been experimenting with traits like this a bit more, and it's all kind of evil as it all depends on what order you "verify" constraints in. Ideally, I feel that the "verify" methods should not solve any constraints and all solving of constraint variables should be done by a separate system which does not depend on the order of constraints, though this would need a bigger refactor.
> 
> If anyone has had any ideas on this in general, it would be good to know.

I think we need to provide an actual order on the traits yes.
This is something that is also broken in MLIR btw, and I am not completely sure on how to fix this in general

**Comment by superlopuh** [comment]
Temporary solutions have a tendency to become permanent. A week is not a long time, and I would feel better if we could all be happy with this PR before merging it. If you see a better to implement it in the relatively near future, then let's wait a little bit longer.

**Comment by superlopuh** [comment]
Oh I forgot about this PR. I'm not sure we reached a concrete conclusion, it would be good to chat about it now that PLDI rush is over.

**Review by superlopuh — APPROVED** [review]

**Comment by alexarice** [comment]
I've rebased on top of the other changes. Finally think this is ready.

**Review by superlopuh — APPROVED** [review]

**Pull request merged**